src/Config.php

```diff
@@ -27,40 +29,49 @@ public function get(string $key, mixed $default = null): mixed
     /**
      * @throws InvalidConfigException
      */
-    private function getByKeyPath(array $key)
+    private function getByKeyPath(array $key): mixed
     {
-        $file = sprintf('%s/%s.php', $this->directory, $key[0]);
+        $subDirectory = array_shift($key);
 
-        if (!file_exists($file)) {
+        $file = realpath("{$this->directory}/{$subDirectory}.php");
+        if (!$file) {
             return null;
         }
 
-        $data = include $file;
-
-        if (!is_array($data)) {
-            throw new InvalidConfigException(sprintf('Config at %s should return array', $file));
+        $data = $this->load($file);
+        foreach ($key as $part) {
+            if (!isset($data[$part])) {
+                return null;
+            }
+            $data = $data[$part];
         }
 
-        array_shift($key);
-
-        return $this->getFromArray($key, $data);
+        return $data;
     }
 
-    private function getFromArray(array $key, array $data): mixed
+    private function load(string $filepath): array
     {
-        if (!count($key)) {
-            return $data;
+        static $loader;
```

Comment:
Having access to `$this` in configs isn't really an issue. The overhead of the closure also makes it considerably slower.

Apart from all that, `(static fn () => include $file)()` does the trick. This is overcomplicated.

**Using a static closure**
```
Took 1.0312781333923
Took 1.0359179973602
Took 1.0410721302032
Took 1.0408849716187
Took 1.0383479595184
Took 1.0358250141144
Took 1.0411012172699
Took 1.0565900802612
Took 1.0560729503632
Took 1.0729780197144
Took 1.0831758975983
Took 1.0484161376953
Took 1.0668151378632
Took 1.0546669960022
Took 1.0482711791992
Took 1.0405220985413
Took 1.039852142334
Took 1.040137052536
Took 1.0360450744629
Took 1.0269370079041
Took 1.027312040329
Took 1.0300810337067
Took 1.0323510169983
Took 1.0355000495911
Took 1.0304670333862
```

**Not doing that**
```
Took 0.96081805229187
Took 0.96303701400757
Took 0.9594190120697
Took 0.98414301872253
Took 0.96710991859436
Took 0.96612501144409
Took 0.96623802185059
Took 0.96417593955994
Took 0.98355102539062
Took 0.98978805541992
Took 0.99073100090027
Took 0.98940014839172
Took 0.98652791976929
Took 0.98738694190979
Took 1.0424950122833
Took 0.98071599006653
Took 0.98969912528992
Took 0.99416089057922
Took 0.98369598388672
Took 0.99504899978638
Took 0.99464201927185
Took 0.98857021331787
Took 0.98969101905823
Took 0.99477291107178
Took 0.99033379554749
```